packages/run-protocol/src/vpool-xyk-amm/pool.js

```diff
@@ -43,210 +43,257 @@ export const makeAddPool = (
-
-        // To calculate liquidity, we'll need to calculate alpha from the primary
-        // token's value before, and the value that will be added to the pool
-        const secondaryRequired = AmountMath.make(
-          secondaryBrand,
-          calcSecondaryRequired(
-            centralIn.value,
-            centralPoolAmount.value,
-            secondaryPoolAmount.value,
-            secondaryIn.value,
-          ),
-        );
+  const addLiquidityActualToState = (
+    state,
+    pool,
+    zcfSeat,
+    secondaryAmount,
+    poolCentralAmount,
+    feeSeat,
+  ) => {
+    const { poolSeat, liquidityBrand, liquidityZcfMint, updater } = state;
+
+    // addLiquidity can't be called until the pool has been created. We
+    // verify that the asset is NAT before creating a pool.
+
+    const liquidityValueOut = calcLiqValueToMint(
+      state.liqTokenSupply,
+      zcfSeat.getStagedAllocation().Central.value,
+      poolCentralAmount.value,
+    );
 
-        // Central was specified precisely so offer must provide enough secondary.
-        assert(
-          AmountMath.isGTE(secondaryIn, secondaryRequired),
-          'insufficient Secondary deposited',
-        );
+    const liquidityAmountOut = AmountMath.make(
+      liquidityBrand,
+      liquidityValueOut,
+    );
+    liquidityZcfMint.mintGains(
+      harden({ Liquidity: liquidityAmountOut }),
+      poolSeat,
+    );
+    state.liqTokenSupply += liquidityValueOut;
+
+    poolSeat.incrementBy(
+      zcfSeat.decrementBy(
+        harden({
+          Central: zcfSeat.getStagedAllocation().Central,
+          Secondary: secondaryAmount,
+        }),
+      ),
+    );
 
-        return addLiquidityActual(pool, zcfSeat, secondaryRequired, centralIn);
-      },
-      removeLiquidity: userSeat => {
-        const liquidityIn = userSeat.getAmountAllocated(
-          'Liquidity',
-          liquidityBrand,
-        );
-        const liquidityValueIn = liquidityIn.value;
-        assert(isNatValue(liquidityValueIn), 'User Liquidity');
-        const centralTokenAmountOut = AmountMath.make(
-          centralBrand,
-          calcValueToRemove(
-            liqTokenSupply,
-            pool.getCentralAmount().value,
-            liquidityValueIn,
-          ),
-        );
+    zcfSeat.incrementBy(
+      poolSeat.decrementBy(harden({ Liquidity: liquidityAmountOut })),
+    );
+    if (feeSeat) {
+      zcf.reallocate(poolSeat, zcfSeat, feeSeat);
+    } else {
+      zcf.reallocate(poolSeat, zcfSeat);
+    }
+    zcfSeat.exit();
+    updateUpdaterState(updater, pool);
+    return 'Added liquidity.';
+  };
 
-        const tokenKeywordAmountOut = AmountMath.make(
-          secondaryBrand,
-          calcValueToRemove(
-            liqTokenSupply,
-            pool.getSecondaryAmount().value,
-            liquidityValueIn,
-          ),
-        );
+  const makePool = defineKind(
+    'pool',
+    (liquidityZcfMint, poolSeat, secondaryBrand) => {
```

Comment:
I'm trying to understand `defineKind`.  How does it know what parameters to provide to the `init` function? (i.e. If the pool doesn't declare them somewhere, what rules produce the parameters?)

Reply:
`defineKind` doesn't "know" the parameters to `init`.  The maker function calls `init` with whatever parameters _it_ was given.